### ISSUE TITLE
OOM copying a file to memory in Blazor

### ISSUE THREAD

**Issue by verdie-g** [dotnet/aspnetcore#55458]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

In a Blazor Wasm app, copying a 1GB file to memory will cause a OOM exception.

I'm aware that, according to https://learn.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-8.0#file-uploads, copying to file in memory is a bad practice. In my exact use-case, I'm not directly copying the bytes, but I'm parsing a very large nettrace file for this project https://verdie-g.github.io/dotnet-event-viewer.

### Expected Behavior

I would expect the browser to allow allocating 1GB on a 16GB machine.

### Steps To Reproduce

1. `dotnet new blazorwasm -n BlazorTest`
2. Replace the content of `Home.razor` with
```
@page "/"

<PageTitle>Home</PageTitle>

<h1>Hello, world!</h1>

<InputFile OnChange="LoadFiles" />

@code {
    private async Task LoadFiles(InputFileChangeEventArgs e)
    {
        List<byte[]> list = new();
        await using var s = e.File.OpenReadStream(maxAllowedSize: long.MaxValue);
        while (true)
        {
            byte[] buffer;
            int read;
            try
            {
                buffer = new byte[64 * 1024];
                read = await s.ReadAsync(buffer);
            }
            catch
            {
                Console.WriteLine("Read " + list.Sum(x => x.Length));
                throw;
            }

            if (read == 0)
            {
                break;
            }

            list.Add(buffer);
        }

        Console.WriteLine("Done");
    }
}
3. `cd BlazorTest && dotnet run`
4. Navigate to the blazor app
5. Upload a 1GB file (create for example with `File.WriteAllBytes("1gb.bin", new byte[1_000_000_000])`
6. Wait
7. Observe the exception in the console


### Exceptions (if any)

When the OOM occurs in `BrowserFileStream.ReadAsync`
```
Microsoft.JSInterop.JSException: Out of memory
   at System.Runtime.InteropServices.JavaScript.JSMarshalerArgument.ToManaged(Byte[]& value)
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.__Wrapper_ReceiveByteArrayFromJS_980098407(JSMarshalerArgument* __arguments_buffer)
Error: Out of memory
    at Jn (http://localhost:5140/_framework/dotnet.runtime.8.0.2.io1j2tel8l.js:3:31614)
    at kr (http://localhost:5140/_framework/dotnet.runtime.8.0.2.io1j2tel8l.js:3:35529)
    at Object.<anonymous> (http://localhost:5140/_framework/dotnet.runtime.8.0.2.io1j2tel8l.js:3:180123)
    at Object.sendByteArray (http://localhost:5140/_framework/blazor.webassembly.js:1:45278)
    at Array.R (http://localhost:5140/_framework/blazor.webassembly.js:1:6844)
    at JSON.stringify (<anonymous>)
    at _ (http://localhost:5140/_framework/blazor.webassembly.js:1:6695)
    at http://localhost:5140/_framework/blazor.webassembly.js:1:2907
   at Microsoft.JSInterop.JSRuntime.<InvokeAsync>d__16`1[[System.Byte[], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at Microsoft.AspNetCore.Components.PullFromJSDataStream.RequestDataFromJSAsync(Int32 numBytesToRead)
   at Microsoft.AspNetCore.Components.PullFromJSDataStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Components.Forms.BrowserFileStream.CopyFileDataIntoBuffer(Memory`1 destination, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Components.Forms.BrowserFileStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at BlazorTest.Pages.Home.LoadFiles(InputFileChangeEventArgs e) in C:\Users\grego\Documents\BlazorTest\Pages\Home.razor:line 26
   at BlazorTest.Pages.Home.LoadFiles(InputFileChangeEventArgs e) in C:\Users\grego\Documents\BlazorTest\Pages\Home.razor:line 37
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
    at b.endInvokeDotNetFromJS (http://localhost:5140/_framework/blazor.webassembly.js:1:3136)
    at Object.gn [as endInvokeDotNetFromJS] (http://localhost:5140/_framework/blazor.webassembly.js:1:58943)
    at http://localhost:5140/_framework/dotnet.runtime.8.0.2.io1j2tel8l.js:3:177853
    at Ul (http://localhost:5140/_framework/dotnet.runtime.8.0.2.io1j2tel8l.js:3:178687)
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[350]:0x1faef
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[246]:0x1bf8c
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[239]:0xf173
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[307]:0x1e7e5
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[328]:0x1efdb
    at http://localhost:5140/_framework/dotnet.native.wasm:wasm-function[218]:0xcfed
```
but it can also occur in the byte array allocation. In the case it's a System.OOM exception.

### .NET Version

9.0.100-preview.3.24204.13

### Anything else?

_No response_

**Comment by mkArtakMSFT** [comment]
@lewing do you know what may prevent a file size of 1GB be a problem? Any limitations in the runtime layer?

@verdie-g what browser are you using?

**Comment by verdie-g** [comment]
> what browser are you using?

Google Chrome.

**Comment by lewing** [comment]
The wasm address space for the MVP is 32 bits, that means that all of your assemblies, your runtime, any allocations, and all I/O buffering needs to fit into 2GB.  To process large nettrace files you will likely need to parse the nettrace file incrementally and stay within the 2GB rss for Wasm data. There are wasm64 extensions that remove this limit but they are not broadly implemented at this time so we'd love to hear more about your use case.

**Comment by verdie-g** [comment]
Ah I should have read more about wasm before starting this project :(

The project is https://verdie-g.github.io/dotnet-event-viewer. A nettrace analyzer tool that could be an alternative to Perfview. Nettrace files can be quite large (e.g. a 20 minutes trace of the GCAllocationTick of one my service weights 10 GB). When, a nettrace is loaded on that tool, it's indeed parsed incrementally, but the thing is that decompressing the data can do a 5x on the size. So obviously that 10GB trace wouldn't even fit in memory but that tool would still cover most of the use-cases if it could use all the available memory.

**Comment by verdie-g** [comment]
Also I'm surprised the 2GB limit is reached there. The file is 1GB but what else could use the other GB?

**Comment by verdie-g** [comment]
> stay within the 2GB rss for Wasm data

If wasm uses 32-bits addressing wouldn't the limit be 4 GiB ? t🤔

**Comment by verdie-g** [comment]
Apologies for the spam but I noticed another issue:
1. Reuse the same code snippet as https://github.com/dotnet/aspnetcore/issues/55458#issue-2273968825
2. Upload a 1 GB file
3. Wait for the OOM
4. Upload a 50 MB
5. Instant OOM

In a non-wasm environment I would expect the GC to be super aggressive to collect objects when the available memory is low. I'm concerned this is not the case here.
